File: medium/add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */
class Solution {
	
	func addTwoNumbers(_ l1: ListNode?, _ l2: ListNode?) -> ListNode? {
		let ln: ListNode? = ListNode(0)
		var listNode: ListNode? = ln
		
		var l1 = l1
		var l2 = l2
		
		while true {
			
			var value = 0
			
			if let l1 = l1 {
				value += l1.val
			}
			
			if let l2 = l2 {
				value += l2.val
			}
			
			if l1 != nil {
				l1 = l1?.next
			}
			
			if l2 != nil {
				l2 = l2?.next
			}
			
			if value >= 10 {
				listNode?.val += value - 10
				listNode?.next = ListNode(1)
			} else {
				listNode?.val += value
				
				if (listNode?.val)! >= 10 {
					listNode?.val = (listNode?.val)! - 10
					listNode?.next = ListNode(1)
				} else if l1 != nil || l2 != nil {
					listNode?.next = ListNode(0)
				}
			}
			
			if listNode?.next != nil {
				listNode = listNode?.next
			}
			
			if l1 == nil && l2 == nil {
				break
			}
		}
		return ln
	}
}

File: medium/add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        s = None
        node = None
        r = 0
        while l1 != None or l2 != None:
            left = 0 if l1 == None else l1.val
            right = 0 if l2 == None else l2.val
            r = left + right + r
            newNode = ListNode(r % 10)
            r = r // 10
            if s == None:
                s = newNode
                node = s
            else:
                node.next = newNode
                node = node.next
            l1 = None if l1 == None else l1.next
            l2 = None if l2 == None else l2.next
        
        if r > 0:
            node.next = ListNode(r)
        return s
            
